### PR TITLE
Use scipy.sparse array datastructure

### PR DESCRIPTION
**Pull request by rossbar**
This is a currently in-progress experiment to see what it would take to adopt the sparse **array** interface proposed in scipy/scipy#14822. Supporting array semantics for sparse data structures will in the long run be a huge improvement for NetworkX IMO. Both dense arrays (i.e. numpy arrays) and sparse arrays are used extensively throughout, and a decent amount of code is dedicated to implementing something for one or the other data structure. The ability to have these data structures behave the same and use them interchangeable would be excellent! Of course, scipy/scipy#14822 doesn't get all the way there yet, but it is a step in the right direction. This PR is to help flesh out the work there and to help start to frame the type of improvements we'd see in NetworkX if such a feature were to be added upstream.

Note also that this is very much a work in progress - the scipy.sparse array API itself is going to be changing a lot, so a lot of the changes here (for example, wrapping objects created with `spdiags` in `sparse.csr_array` calls) is subject to change.

### CODE REVIEW

File: networkx/algorithms/link_analysis/hits_alg.py
Comment (jarrodmillman) [inline networkx/algorithms/link_analysis/hits_alg.py:368]: I like that you are using A for the adjacency matrix above.  Maybe we should keep that convention.  Then I would replace
 ```
A = M.T @ M  # authority matrix	
```
with
```
ATA = A.T @ A # authority matrix	
```
Similarly,
```
AAT = A @ A.T # hub matrix	
```

File: networkx/algorithms/link_analysis/hits_alg.py
Comment (dschult) [inline networkx/algorithms/link_analysis/hits_alg.py:373]: It looks like if nstart is not None we set x=1/n and then go into the if statement where x gets set (and we wasted the setup time).
Would it be better to leave the if clause for when nstart is None and remove line 369 instead of removing the if clause?

**Review by dschult — COMMENTED** [review]
Looks good so far. I like the notation of places where we should sparsify or not densify.
Just one comment about whether setting x should occur in an if statement or not.

Nice!

**Comment by rossbar** [comment]
The pre-release testing workflows (which pick up the scipy 1.8 rc) are now passing :tada:

**Comment by jarrodmillman** [comment]
@rossbar @dschult @stefanv @MridulS  I would like to merge this as soon as possible.  We still need to do this:

- [ ] change `scipy==1.8.0rc2` to `scipy>=1.8.0` (once 1.8 is released) in `requirements/default.txt`

Is there anything else that needs to be done before merging this?  Once this is merged, what are the next steps?  Are there more cases where we should add a `FutureWarning` (e.g., #5219)?

**Comment by rossbar** [comment]
> Is there anything else that needs to be done before merging this?

Yes I still need to add FutureWarnings where the functions outputs will change. I will do so ASAP!

**Comment by rossbar** [comment]
There are now `FutureWarnings` for all functions that currently return scipy.sparse matrices to indicate that they will return scipy.sparse arrays instead in NX 3.0.

One remaining decision point is what to do about the main graph->sparse conversion function, i.e. `nx.to_scipy_sparse_matrix`. This function currently returns a sparse matrix, but all of the changes in this PR that update to array semantics expect this to return a scipy.sparse array. One solution would be to add a new conversion function `nx.to_scipy_sparse_array` that is explicit about the fact that it returns sparse arrays rather than matrices. That way, we can use `to_scipy_sparse_array` internally and start recommending it as the preferred conversion function moving forward, while also avoiding breaking users without warning by altering what `to_scipy_sparse_matrix` returns. LMK what you think about this approach!

**Comment by rossbar** [comment]
I think this PR is now in a place where it's ready for review/discussion so I've removed the WIP label.

**Review by dschult — COMMENTED** [review]
The main questions I have are about `tests/test_convert_scipy.py` where there are tests of `to_scipy_sparse_array`, but none for `from_scipy_sparse_array`.  I'm guessing that the existing tests for `from_scipy_sparse_matrix` can just be switched over to the array api. Do we want to do that now? Do we want to duplicate it now (convert one copy to array) and then remove the version for matrix later?

Otherwise this look good.  I'm assuming that the sparse array itself is being tested elsewhere. So we're just testing the NetworkX usage of this data structure.

**Comment by rossbar** [comment]
> The main questions I have are about tests/test_convert_scipy.py where there are tests of to_scipy_sparse_array, but none for from_scipy_sparse_array

This was because I hadn't added `from_scipy_sparse_array` yet - done as of 4a92d07.

> I'm guessing that the existing tests for from_scipy_sparse_matrix can just be switched over to the array api.

Correct, this was also done in 4a92d07. In fact, `from_scipy_sparse_array` should work fine with either sparse `array` or `matrices` without any changing at all.

**Review by jarrodmillman — APPROVED** [review]

**Comment by jarrodmillman** [comment]
@MridulS @dschult I think we should merge this before scipy 1.8 is released.  Then we can merge #5262.  Before releasing networkx 2.7, we will need to update the requirements file.  Since this is a big change it would be good to have it live in the main branch for a bit (potentially more people we see it and possibly test it for us).

File: networkx/convert_matrix.py
Comment (MridulS) [inline networkx/convert_matrix.py:990]: Should we also put in a `FutureWarning` here? Nudge people to use `to_scipy_sparse_array` instead of `to_scipy_sparse_matrix`. (Not sure if this has been already discussed 😅 )

File: networkx/convert_matrix.py
Comment (MridulS) [inline networkx/convert_matrix.py:976]: We should probably keep this as `to_scipy_sparse_matrix`, it's the documentation of `to_scipy_sparse_matrix`.

**Review by MridulS — APPROVED** [review]
Looks good to me! Thanks @rossbar!

File: networkx/convert_matrix.py
Comment (MridulS) [reply networkx/convert_matrix.py:990]: Ah wait scratch that, this is taken care in https://github.com/networkx/networkx/pull/5262/commits/d9b8c7c712485735c3de211160a00f0e43a4204c

Comment (rossbar) [reply networkx/convert_matrix.py:990]: I originally had this as just a FutureWarning, but after last week's meeting decided to switch to the more "aggressive" DeprecationWarning, which I then did in a separate PR to follow this one: #5262 . So there will be a warning, but feedback on whether it should be a deprecation or just a more gentle nudge would still be appreciated :)

**Comment by dschult** [comment]
Wow -- there is a lot in here.... :}  LGTM...
Thanks @rossbar!!

File: networkx/convert_matrix.py
Comment (rossbar) [reply networkx/convert_matrix.py]: Good catch! I had the opposite problem in the `from_scipy_sparse_array` exmaples as well. Updated examples in 252108f

**Comment by MridulS** [comment]
I think we can merge this in now, we can further iterate on this in https://github.com/networkx/networkx/pull/5262

Are there any other blockers @rossbar ?

**Comment by rossbar** [comment]
> Are there any other blockers

Nope, not at this point. We're still waiting on the official release of scipy 1.8, but the dependency updates can be handled in a separate PR once it's out.

File: examples/algorithms/plot_subgraphs.py
Comment (stefanv) [inline examples/algorithms/plot_subgraphs.py:1]: Related to this PR?

File: examples/drawing/plot_tsp.py
Comment (stefanv) [inline examples/drawing/plot_tsp.py:1]: Related to this PR?

**Comment by MridulS** [comment]
I'll go ahead and merge this then 😄

File: networkx/convert_matrix.py
Comment (stefanv) [inline networkx/convert_matrix.py:1020]: Missing returns section

File: networkx/convert_matrix.py
Comment (stefanv) [inline networkx/convert_matrix.py:999]: Nitpick, but ideally these fit on one line. We also typically use "Create" instead of "Creates", i.e., something like:

"Create a graph from a sparse adjacency matrix."

File: networkx/linalg/bethehessianmatrix.py
Comment (stefanv) [inline networkx/linalg/bethehessianmatrix.py:38]: I like this level of specificity.  Should the same be used above, instead of simply "SciPy sparse matrix"?  Or were you keeping your options open there, in case it might change in the future?

File: networkx/linalg/bethehessianmatrix.py
Comment (stefanv) [inline networkx/linalg/bethehessianmatrix.py:74]: This is a nicely descriptive TODO.  Above there is one that simply says "TODO: csr_array"

**Review by stefanv — APPROVED** [review]
A thorough piece of work, thank you @rossbar.  Some tiny pieces of feedback, to use or lose as you choose.

**Comment by stefanv** [comment]
Ah, perfect timing as usual :joy:

**Comment by MridulS** [comment]
Woopsie, sorry about that 😅

File: examples/algorithms/plot_subgraphs.py
Comment (rossbar) [reply examples/algorithms/plot_subgraphs.py:1]: Yeah I had to merge with `main` at least twice during the development to resolve conflicts and the rebase was pretty complicated so I was lazy and just merged, relying on the final squash to clean up the history "/. So no - these changes aren't directly related.

**Pull request merged**
